### ISSUE TITLE
TypeScript generation to support grpc-node

### ISSUE THREAD

**Issue by MOZGIII** [dcodeIO/protobuf.js#1017]
I'm trying to use protobuf.js with [grpc-node native core](https://github.com/grpc/grpc-node/tree/master/packages/grpc-native-core), and want to have a proper TypeScript definitions.

I have already mentioned this at #1007, but I'm really struggling with incompatibilities of the generated TypeScript types and the grpc implementation. Decided it's worth a separate issue here.

So, the problem is, TypeScript generated code does not reflect the existence of the streams in the rpc definition.

Example:

```protobuf
syntax = "proto3";

package example;

service StreamExample {
  rpc NoStreams(Request) returns (Response) {}
  rpc StreamingResponse(Request) returns (stream Response) {}
  rpc StreamingRequest(stream Request) returns (Response) {}
  rpc StreamingBoth(stream Request) returns (stream Response) {}
}

message Request {
  string a = 1;
  string b = 2;
}

message Response {
  string c = 1;
  string d = 2;
}
```

```typescript
export namespace example {
    class StreamExample extends $protobuf.rpc.Service {
        public noStreams(request: example.IRequest, callback: example.StreamExample.NoStreamsCallback): void;
        public noStreams(request: example.IRequest): Promise<example.Response>;

        public streamingResponse(request: example.IRequest, callback: example.StreamExample.StreamingResponseCallback): void;
        public streamingResponse(request: example.IRequest): Promise<example.Response>;

        public streamingRequest(request: example.IRequest, callback: example.StreamExample.StreamingRequestCallback): void;
        public streamingRequest(request: example.IRequest): Promise<example.Response>;

        public streamingBoth(request: example.IRequest, callback: example.StreamExample.StreamingBothCallback): void;
        public streamingBoth(request: example.IRequest): Promise<example.Response>;
    }
    namespace StreamExample {
        type NoStreamsCallback = (error: (Error|null), response?: example.Response) => void;
        type StreamingResponseCallback = (error: (Error|null), response?: example.Response) => void;
        type StreamingRequestCallback = (error: (Error|null), response?: example.Response) => void;
        type StreamingBothCallback = (error: (Error|null), response?: example.Response) => void;
    }
}
```
(comments and some code parts are removed for clarity)

As seen from the example, there definitions for function calls in TypeScript are the same despite them being different in the protobuf definition.

I'd like to discuss and, probably, correctly implement the extensive support for protobuf format and compatibility with grpc-node.

This is a starting point. I'll publish a sample repo soon with the examples to we can have something to play with.

**Comment by coding2012** [comment]
Note that Promises are not really allowed to have multiple values (they officially end after one value). But RXJS Observables seem to be a good fit for multiple values... they definitely are designed for streams in this sense.

**Comment by MOZGIII** [comment]
There's been quite a discussion at https://github.com/grpc/grpc-node/issues/528 about this recently.
Anyone interested, please check that out.

**Comment by MOZGIII** [comment]
TL;DR: a dedicated support is needed from protobuf.js, in particular we need better understanding and type definitions for the format that `fromObject`/`toObject` operate on:

https://github.com/dcodeIO/protobuf.js/blob/69623a91c1e4a99d5210b5295a9e5b39d9517554/tests/data/rpc.d.ts#L27-L28

The type definitions for those two pretty much allow anything, while it practice there are concrete requirements to the input, and the output has a particular shape.

**Comment by trajano** [comment]
I was able to successfully create a server in TypeScript with protobufjs along with a few additions.  https://stackoverflow.com/a/64778825/242042

The protobufjs generates the request and response callbacks properly the most part.  However, I had to create the "service" types as follows

```
interface IArtifactUpload {
  signedUrlPutObject(call: ServerUnaryCall<IUploadRequest>, callback: ArtifactUpload.SignedUrlPutObjectCallback): void;
}
interface ServerDefinition extends GrpcObject {
  service: GrpcObject & ServiceDefinition<IArtifactUpload>;
}
interface ServerPackage extends GrpcObject {
  [name: string]: ServerDefinition
}
const protoDescriptor = loadPackageDefinition(packageDefinition) as ServerPackage;

```

**Comment by trajano** [comment]
I've found a way of doing a slightly cleaner typescript and made some alterations to pretend I am crazy enough to put more than one `service` definition in a single proto file.

These are the common interfaces:

```
interface ServerDefinition<W> extends GrpcObject {
  service: GrpcObject & ServiceDefinition<W>;
}
interface ServerPackage<W> extends GrpcObject {
  [name: string]: ServerDefinition<W>
}
```
And here are my application specific code which I think can be generated with a tool.
```
interface IArtifactUpload {
  signedUrlPutObject: handleUnaryCall<IUploadRequest, ISignedUrlPutObjectResponse>;
}
interface IArtifactDownload {
  foo: handleUnaryCall<IUploadRequest, ISignedUrlPutObjectResponse>;
}
type CustomApi = IArtifactUpload | IArtifactDownload;
```
Then:
```
const protoDescriptor = loadPackageDefinition(packageDefinition) as ServerPackage<CustomApi>;
```

Then the server and implementations:
```
const server = new Server();
server.addService(protoDescriptor.ArtifactUpload.service as ServiceDefinition<IArtifactUpload>, {
  signedUrlPutObject(call, callback) {
    callback(null, SignedUrlPutObjectResponse.create({ reply: "hello " + call.request.message }));
  }
});
server.addService(protoDescriptor.ArtifactUpload.service as ServiceDefinition<IArtifactDownload>, {
  foo(call, callback) {
    callback(null, SignedUrlPutObjectResponse.create({ reply: "hello " + call.request.message }));
  }

});
server.bind('0.0.0.0:50051', ServerCredentials.createInsecure());
server.start();
```

**Comment by dcodeIO** [comment]
Closing this issue because it has been inactive for more than two years and the project has changed substantially.

If this is still relevant, please open a fresh issue with current reproduction steps, version details, and any updated context.

Please also consider that maintaining protobuf.js takes time and effort. If the project is valuable to you or your organization, sponsorship is appreciated.

Thanks!
